pcsx2/IPC.cpp

```diff
@@ -0,0 +1,378 @@
+	// multiple emulators running and make this a folder
+#ifdef __APPLE__
+	char* runtime_dir = std::getenv("TMPDIR");
+#else
+	char* runtime_dir = std::getenv("XDG_RUNTIME_DIR");
+#endif
+	// fallback in case macOS or other OSes don't implement the XDG base
+	// spec
+	if (runtime_dir == NULL)
+		m_socket_name = (char*)"/tmp/pcsx2.sock";
+	else
+		m_socket_name = strcat(runtime_dir, "/pcsx2.sock");
+
+	struct sockaddr_un server;
+
+	m_sock = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (m_sock < 0)
+	{
+		Console.WriteLn(Color_Red, "IPC: Cannot open socket! Shutting down...");
+		return;
+	}
+	server.sun_family = AF_UNIX;
+	strcpy(server.sun_path, m_socket_name);
+
+	// we unlink the socket so that when releasing this thread the socket gets
+	// freed even if we didn't close correctly the loop
+	unlink(m_socket_name);
+	if (bind(m_sock, (struct sockaddr*)&server, sizeof(struct sockaddr_un)))
+	{
+		Console.WriteLn(Color_Red, "IPC: Error while binding to socket! Shutting down...");
+		return;
+	}
+#endif
+
+	// maximum queue of 4096 commands before refusing, approximated to the
+	// nearest legal value. We do not use SOMAXCONN as windows have this idea
+	// that a "reasonable" value is 5, which is not.
+	listen(m_sock, 4096);
+
+	// we save a handle of the main vm object
+	m_vm = vm;
+
+	// we start the thread
+	Start();
+}
+
+void SocketIPC::ExecuteTaskInThread()
+{
+	m_end = false;
+
+	// we allocate once buffers to not have to do mallocs for each IPC
+	// request, as malloc is expansive when we optimize for µs.
+	m_ret_buffer = new char[MAX_IPC_RETURN_SIZE];
+	m_ipc_buffer = new char[MAX_IPC_SIZE];
+	while (true)
+	{
+		m_msgsock = accept(m_sock, 0, 0);
+		if (m_msgsock == -1)
+		{
+			// everything else is non recoverable in our scope
+			// we also mark as recoverable socket errors where it would block a
+			// non blocking socket, even though our socket is blocking, in case
+			// we ever have to implement a non blocking socket.
+#ifdef _WIN32
+			int errno_w = WSAGetLastError();
+			if (!(errno_w == WSAECONNRESET || errno_w == WSAEINTR || errno_w == WSAEINPROGRESS || errno_w == WSAEMFILE || errno_w == WSAEWOULDBLOCK))
+			{
+#else
+			if (!(errno == ECONNABORTED || errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK))
+			{
+#endif
+				fprintf(stderr, "IPC: An unrecoverable error happened! Shutting down...\n");
+				m_end = true;
+				break;
+			}
+		}
+		else
+		{
+
+#ifdef _WIN32
+			// socket timeout
+			DWORD tv = 10 * 1000; // 10 seconds
+#else
+			// socket timeout
+			struct timeval tv;
+			tv.tv_sec = 10;
+			tv.tv_usec = 0;
+#endif
+			setsockopt(m_msgsock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&tv, sizeof tv);
+			setsockopt(m_msgsock, SOL_SOCKET, SO_SNDTIMEO, (const char*)&tv, sizeof tv);
+
+
+			// either int or ssize_t depending on the platform, so we have to
+			// use a bunch of auto
+			auto receive_length = read_portable(m_msgsock, m_ipc_buffer, MAX_IPC_SIZE);
+			if (receive_length > 0)
```

Comment:
I think you can avoid some of initialization by just starting with `receive_length = 0` and `end_length = 4`, then you can skip the check here and just let everything be read and checked by the while loop